### PR TITLE
Add Functionality for Adding Assignments & Materials and Modifying Course Parameters

### PR DESCRIPTION
**Pull request by FlorianDinant**
This PR introduces two key components: the AddAssignment component for creating new assignments in a course and the CourseParameters component for modifying course details like name, description, credits, and section. These components are designed to provide a flexible and user-friendly interface for users managing course assignments and parameters. The buttons used to edit the course are accessible (and displayed) for teacher users only.

### `SelectActions` Component:
   - When clicking on the "+" button at the bottom right corner of the course screen 2 buttons are displayed : 
      - Add Assignment : which allows to open the modal to add an assignment to the course
      - Add Material : which allows to open the modal to add a material to the course
### `AddAssignment` Component:
   - `Name` Input: Users can input the name of the assignment with the ability to trigger changes dynamically.
   - Assignment `Type` Button: Users can switch between assignment types (quiz or submission) using a button, with dynamic color updates to reflect the selection.
   - `Due Date` and `Time Pickers`: Users can pick a due date and time via DateTimePicker for better flexibility. The date and time pickers are displayed only when needed, ensuring a clean user interface.
   - Submission Button: A button is provided for submitting the assignment once all fields are filled, ensuring that data is properly validated before submission.
### `AddMaterial` Component:
   - `Title` Input: Users can input the title of the Material with the ability to trigger changes dynamically.
   - `Description` Input: Users can input the description of the assignment with the ability to trigger changes dynamically.
   - `From Date` and `FromTime` Inputs:  Users can pick a start date and time via DateTimePicker for better flexibility. The date and time pickers are also displayed only when needed, ensuring a clean user interface.
   - - `To Date` and `To Time` Inputs:  Same as above but for ending date
   - Submission Button: A button is provided for submitting the material once all fields are filled, ensuring that data is properly validated before submission.
### `CourseParameters` Component:
   - This component is only displayed if the user (teacher) clics on the parameter icon at the top right corner of the course page screen
   - Dynamic Course Data Editing: Allows modification of course parameters like name, description, section, and credits.
   - `Section` Selection: Users can change the section (e.g., IN, SC, MA) using a button that cycles through available sections with color-coded buttons to enhance usability.
   - `Credits` Adjustment: Users can increase or decrease the course credits between 0 and 30 using + and - buttons, with the current value displayed.
   - Course Details Input: Includes inputs for the course’s name and description with FancyTextInput for a smooth user experience. The inputs are validated to ensure data integrity before saving.

### 1. SelectActions Component:
   - **Functionality**:
      - Positioned at the bottom right of the course screen, a “`+`” button appears when clicked. This triggers the display of two action buttons:
      - **Add Assignment**: Opens the AddAssignment modal for adding assignments to the course.
      - **Add Material**: Opens the AddMaterial modal for adding materials to the course.
   - **UI Design**:
      - Floating action button (`+`) with prominent, accessible placement to enhance ease of use.
      - Smooth transitions when expanding the action buttons to maintain clean, user-friendly interactions.

### 2. AddAssignment Component:
   - **Inputs**:
      - Assignment `Name`: Text input with dynamic value changes, allowing the user to enter and modify the assignment name.
      - Assignment `Type` Button: Toggle button to switch between Quiz and Submission. Buttons are dynamically color-coded to visually reflect the selected type.
      - `Due Date and Time`: Date and Time pickers allow users to select the assignment’s due date and time. These fields are only displayed when needed to minimize screen clutter.
   - **Validation**:
      - Ensure that the assignment name is non-empty.
      - Ensure that the date and time are valid and future-dated.
      - Submission Button: A submission button that remains disabled until all required fields are filled. Once active, it validates inputs and submits the assignment details.

### 3. AddMaterial Component:
   - **Inputs**:
      - Material Title: Input field for the material’s title with dynamic changes.
      - Description: Input field for entering a detailed description of the material.
      - From Date and From Time: DateTimePicker for selecting the starting date and time for the material.
      - To Date and To Time: DateTimePicker for selecting the end date and time.
   - **Validation**:
      - Ensure that all fields have valid inputs (e.g., dates should be in the future, description should not be empty).
      - Submission Button: A button that becomes active only when all fields are correctly filled. This triggers the submission process for the material.

### 4. CourseParameters Component:
   - **Visibility**:
      - This component is only visible when the user (teacher) clicks on the parameter icon located at the top right corner of the course page.
   - **Dynamic Editing**:
      - Course `Name` and `Description`: Users can edit the course name and description using text inputs with dynamic state binding.
      - `Section`: Button that allows users to toggle through different course sections (e.g., IN, SC, MA), with color-coded buttons for easy identification.
      - `Credits`: Users can modify the course’s credit value between 0 and 30 by clicking the + and - buttons, and the current credit value is displayed. This ensures clear visibility of the credit value.
   - **Validation**:
      - Validate that the course name is non-empty and that the credit value is within the accepted range (0 to 30).
   - **UI Elements**:
      - FancyTextInput for smooth, visually appealing text editing.
      - Save button to confirm changes.

## UI Enhancements:
   - **Dynamic Feedback**: Interactive inputs with immediate visual feedback (e.g., color changes, validation checks) to improve the user experience.
   - **Visibility Control**: DateTimePickers for both assignment and material components are conditionally rendered, ensuring a clean interface when not needed.
   - **Color Coding**: Key interactive elements (like assignment type buttons and section buttons) are color-coded to indicate their state, helping users understand their current selection.
   - **Responsiveness**: All components are fully responsive and optimized for both mobile and desktop views, ensuring consistent behavior across devices.
   - **Smooth Transitions**: Smooth animations and transitions when the action buttons are displayed and when changing between sections or assignment types.

Notes:
   - **Date-Time Pickers**: I inspired myself when using it from Adamm's todo page. The inclusion of DateTimePickers ensures flexibility in specifying precise deadlines or start/end times for assignments and materials. These are displayed only when needed, reducing screen clutter.
   - **Dynamic Content**: The ability to dynamically modify course parameters, including the course name, description, credits, and section, makes it easy to keep course information up-to-date.
   - **Modal Forms**: Both the AddAssignment and AddMaterial components use modal forms for a streamlined, non-intrusive experience. Modals allow users to focus on completing the task without leaving the current page.

## Testing:
   - THIS PR DOES NOT PROVIDE UI/UNIT TESTS YET (it will in a future PR)
   - For now please test as follows : 
      - Test that edit buttons at the top and bottom right corners are not visible for non-professor user.
      - Ensure that all interactive elements like the date/time pickers and buttons work as expected.
      - Test the assignment submission logic to make sure that only valid inputs allow the submission to be enabled.
      - Ensure that it is not possible to give a credit ECTS value less than 0 or greater than 30.

## Screenshots

Screenshot 2024-11-28 at 23 22 29
Screenshot 2024-11-28 at 23 23 55
Screenshot 2024-11-28 at 23 23 26
Screenshot 2024-11-28 at 23 22 42
Screenshot 2024-11-28 at 23 22 51
Screenshot 2024-11-28 at 23 23 12
Screenshot 2024-11-28 at 23 54 38

This PR is designed to streamline the process of adding assignments and modifying course parameters in a flexible and intuitive manner.

### CODE REVIEW

File: edweiss-app/components/input/FancyButton.tsx
Comment (Gwen678) [inline edweiss-app/components/input/FancyButton.tsx:23]: Why do you use `?` 
Maybe you can add some defaults attributs.

**Review by Gwen678 — APPROVED** [review]
The code is truly massive and impressive! It's excellent work, and I love how everything is structured and works. I don't have much to say because everything looks perfect to me. I've only added a few minor details, but nothing really important. For me, it's perfect!

Great job👏

File: edweiss-app/components/input/FancyButton.tsx
Comment (FlorianDinant) [reply edweiss-app/components/input/FancyButton.tsx:23]: I prefer to leave it like this since outlined, style etc... don't have default values. It does not poses any problem to leave it like this imo.

File: edweiss-app/components/courses/SelectActionsCourse.tsx
Comment (FlorianDinant) [reply edweiss-app/components/courses/SelectActionsCourse.tsx:69]: Done thank you !

File: edweiss-app/app/(app)/(tabs)/index.tsx
Comment (FlorianDinant) [reply edweiss-app/app/(app)/(tabs)/index.tsx:112]: Yes you are right

**Comment by FlorianDinant** [comment]
> The code is truly massive and impressive! It's excellent work, and I love how everything is structured and works. I don't have much to say because everything looks perfect to me. I've only added a few minor details, but nothing really important. For me, it's perfect!
> 
> Great job👏

Thank you for your review and the suggested changes! I am happy you globally agreed on my code. We can now wait for Ferdinand's review before merging and we'll be done!

File: edweiss-app/components/courses/AddAssignment.tsx
Comment (FerdinandGomerChiu) [inline edweiss-app/components/courses/AddAssignment.tsx:186]: For quizzes, we'll have to redirect to the quiz creation page, with the name and due date passed down in the router. The submission logic of quizzes is handled in the creation page.

**Comment by FerdinandGomerChiu** [comment]
This is pretty nitpicky, but you could take a look at sonar cloud analysis and try to fix the little issues it highlights, such as:
image
That being said, I believe you still pass the quality check.

**Review by FerdinandGomerChiu — APPROVED** [review]
Very good job, this is some huge code. It is clearly written, I did not have a lot to say. 
In the future, we'll just have to properly link with quiz and slides creation. Great work!

File: edweiss-app/components/courses/AddAssignment.tsx
Comment (FlorianDinant) [reply edweiss-app/components/courses/AddAssignment.tsx:186]: Yep let's do that on the next sprint!

**Pull request merged**
